core/src/io/anuke/mindustry/world/blocks/power/PowerDiode.java

```diff
@@ -0,0 +1,113 @@
+package io.anuke.mindustry.world.blocks.power;
+
+import io.anuke.arc.Core;
+import io.anuke.arc.graphics.g2d.Draw;
+import io.anuke.arc.graphics.g2d.TextureRegion;
+import io.anuke.arc.util.Eachable;
+import io.anuke.mindustry.entities.traits.BuilderTrait;
+import io.anuke.mindustry.entities.type.TileEntity;
+import io.anuke.mindustry.graphics.Pal;
+import io.anuke.mindustry.ui.Bar;
+import io.anuke.mindustry.ui.Cicon;
+import io.anuke.mindustry.world.Block;
+import io.anuke.mindustry.world.Tile;
+
+public class PowerDiode extends Block{
+
+    protected TextureRegion arrow;
+
+    public PowerDiode(String name){
+        super(name);
+        rotate = true;
+        update = true;
+        solid = true;
+        insulated = true;
+    }
+
+    @Override
+    public void update(Tile tile){
+        super.update(tile);
+
+        DiodeEntity entity = (DiodeEntity) tile.entity;
+
+        if(!entity.connected()) return;
+        if(entity.input() > entity.output()){
+            entity.transfer(entity.graph(entity.back()).getBatteryStored() * (entity.input() - entity.output()) / 2);
+        }
+    }
+
+    @Override
+    public void setBars(){
+        super.setBars();
+
+        bars.add("back", entity -> new Bar("bar.input", Pal.lighterOrange, () -> ((DiodeEntity) entity).input()) );
+        bars.add("front", entity -> new Bar("bar.output", Pal.lighterOrange, () -> ((DiodeEntity) entity).output()) );
+    }
+
+    @Override
+    public void load(){
+        super.load();
+        arrow = Core.atlas.find(name + "-arrow");
+    }
+
+    @Override
+    public void draw(Tile tile){
+        Draw.rect(region, tile.drawx(), tile.drawy(), 0);
+        Draw.rect(arrow, tile.drawx(), tile.drawy(), rotate ? tile.rotation() * 90 : 0);
+    }
+
+    @Override
+    public void drawRequestRegion(BuilderTrait.BuildRequest req, Eachable<BuilderTrait.BuildRequest> list) {
+        TextureRegion reg = icon(Cicon.full);
+        Draw.rect(icon(Cicon.full), req.drawx(), req.drawy(),
+                reg.getWidth() * req.animScale * Draw.scl,
+                reg.getHeight() * req.animScale * Draw.scl,
+                0);
+        Draw.rect(arrow, req.drawx(), req.drawy(),
+                arrow.getWidth() * req.animScale * Draw.scl,
+                arrow.getHeight() * req.animScale * Draw.scl,
+                !rotate ? 0 : req.rotation * 90);
+    }
+
+    @Override
+    public TileEntity newEntity(){
+        return new DiodeEntity();
+    }
+
+    public static class DiodeEntity extends TileEntity{
+
+        public Tile back(){
```

Comment:
Some notes about these methods:
1. The `front()` and `back()` methods seem like utilities that could be more useful in `Tile`, not this entity class. In fact, one of my branches has these methods in it.
2. I am not a fan of making turning everything into a mini-method like this. Trying to figure out what's going on in the main `update()` method is very difficult, considering I need to look through 2-3 different method calls in a different part of the file to figure out what it's actually doing. [This earlier version](https://github.com/Anuken/Mindustry/blob/2cd8fe0af9fcb1f3014e90f054ba52d69830265a/core/src/io/anuke/mindustry/world/blocks/power/PowerDiode.java#L25) seems a lot more readable: I can see what exactly it's doing, without having to untangle what each method's task is. If these methods were used extensively for many different classes, that would be another story, but in this case it seems like the only thing they will ever be used for is one `update()` method.
3. Even if you *needed* new utility methods, it would be a better idea to put them in the block class instead of creating a new `DiodeEntity` class. `TileEntity` classes are only for storing block specific data.

Reply:
1. moved

2. reverted

3. noted